### DISCUSSION TITLE
Details about UncheckedInput

### DISCUSSION THREAD

**Discussion by ymc9** [prisma/prisma#22809] [Q&A] [answered]
### Question

Hi guys,

I roughly understand that the generated UncheckedInput types are for unsafe operations like directly setting fk fields. However, I'm unclear about the detailed rules on when a field gets generated into both regular and unchecked input types, and when only into unchecked ones.

For example, I found that if a `@id` field has an `autoincrement()` default, it's considered unsafe. But other kinds of id fields (like Int without `autoincrement`, or String, etc.) are safe. I'm wondering if there's a specification about the rules. Thanks!

**Comment by ymc9** [comment]
Btw, the reason why I care about this is that we inject into Prisma queries, so need to be sure which form (safe or unsafe) is expected by Prisma.

**Reply by ludralph** [reply]
Hey @ymc9! 👋 

I have shared your question with the team and we'll get back to you soon.

**Reply by ymc9** [reply]
Thank you!

**Reply by ludralph** [reply, marked as answer]
Hi @ymc9 
Thank you for your patience. The rules can be seen below.  You can also peek into the [code](https://github.com/prisma/prisma-engines/blob/79f6f09ba683cc77d70f9ea0cf6177a3e35a2525/query-engine/schema/src/build/mutations/create_one.rs#L110) responsible for the checked/unchecked input generation.

**Checked Input Types:**
**Purpose**: Ensure data integrity by enforcing certain restrictions on fields that can be written directly.
**Restrictions**:

- Excludes autogenerated IDs.
- Excludes relation fields that would create endless nested types.
- Allows writable scalar fields that are not read-only or autogenerated IDs.

Use Cases:

- When data integrity is paramount and direct writes to sensitive fields need to be controlled.
- When preventing potential issues caused by unintentionally modifying related data.

**Unchecked Input Types**:
**Purpose**: Provide more flexibility for writing data, even if it involves potential risks to data integrity.
**Allowed Fields**:

- All writable scalar fields, except those used for linking in nested creates.
- Relation fields that point to other models and are not inlined on the current model.

**Use Cases:**

- When more control over data manipulation is needed, even at the expense of some potential risks.
- When working with complex data structures or relationships that require more direct field access.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by ymc9** [reply]
Thank you, @ludralph, for providing such a detailed description! It helps tremendously!
